src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java

```diff
@@ -0,0 +1,440 @@
+ * this class will omit the flush at t = 5s and trigger a flush at t = 9s. If
+ * ERROR log events keep being generated before a flush occurs, then this class
+ * will definitely trigger a flush at t = 5min based on the hard timeout.
+ * <p></p>
+ * Maintaining these timeouts per log level allows us to flush logs sooner if
+ * more important log levels occur. For instance, we can set smaller timeouts
+ * for ERROR log events compared to DEBUG log events.
+ */
+public abstract class AbstractBufferedRollingFileAppender extends EnhancedAppenderSkeleton
+    implements Flushable
+{
+  protected String currentLogPath = null;
+  protected final TimeSource timeSource;
+
+  // Appender settings, some of which may be set by Log4j through reflection.
+  // For descriptions of the properties, see their setters below.
+  private boolean closeFileOnShutdown = true;
+  private final HashMap<Level, Long> flushHardTimeoutPerLevel = new HashMap<>();
+  private final HashMap<Level, Long> flushSoftTimeoutPerLevel = new HashMap<>();
+  private int timeoutCheckPeriod = 1000;
+
+  private long flushHardTimeoutTimestamp;
+  private long flushSoftTimeoutTimestamp;
+  // The maximum soft timeout allowed. This is used while the app is shutting
+  // down to increase the likelihood of flushing before the app finishes
+  // shutting down.
+  private long flushMaximumSoftTimeout;
+
+  private final BackgroundFlushThread backgroundFlushThread = new BackgroundFlushThread();
+  private final BackgroundSyncThread backgroundSyncThread = new BackgroundSyncThread();
+
+  private boolean activated = false;
+
+  public AbstractBufferedRollingFileAppender () {
+    this(new SystemTimeSource());
+  }
+
+  /**
+   * Constructor
+   * @param timeSource The time source that the appender should use
+   */
+  public AbstractBufferedRollingFileAppender (TimeSource timeSource) {
+    this.timeSource = timeSource;
+
+    // The default flush timeout values below are optimized for high latency
+    // remote persistent storage such as object stores or HDFS
+    flushHardTimeoutPerLevel.put(Level.FATAL, 5L * 60 * 1000 /* 5 min */);
+    flushHardTimeoutPerLevel.put(Level.ERROR, 5L * 60 * 1000 /* 5 min */);
+    flushHardTimeoutPerLevel.put(Level.WARN, 10L * 60 * 1000 /* 10 min */);
+    flushHardTimeoutPerLevel.put(Level.INFO, 30L * 60 * 1000 /* 30 min */);
+    flushHardTimeoutPerLevel.put(Level.DEBUG, 30L * 60 * 1000 /* 30 min */);
+    flushHardTimeoutPerLevel.put(Level.TRACE, 30L * 60 * 1000 /* 30 min */);
+
+    flushSoftTimeoutPerLevel.put(Level.FATAL, 5L * 1000 /* 5 sec */);
+    flushSoftTimeoutPerLevel.put(Level.ERROR, 10L * 1000 /* 10 sec */);
+    flushSoftTimeoutPerLevel.put(Level.WARN, 15L * 1000 /* 15 sec */);
+    flushSoftTimeoutPerLevel.put(Level.INFO, 3L * 60 * 1000 /* 3 min */);
+    flushSoftTimeoutPerLevel.put(Level.DEBUG, 3L * 60 * 1000 /* 3 min */);
+    flushSoftTimeoutPerLevel.put(Level.TRACE, 3L * 60 * 1000 /* 3 min */);
```

Comment:
Is it possible there's no logging level? Like client abuse the system and log with logger.log() instead of logger.info, etc.

Reply:
I do not believe log4j1 supports logging without specifying the logging level. All of the interfaces for `logger.log()` require a mandatory Priority parameter. It is possible that user defines a custom log level, but that is extremely rare and we should probably check for that in `setFlushSoftTimeoutsInSeconds` and `setFlushHardTimeoutsInMinutes`

Reply:
Yes, that's what I mean, to catch such edge case and add a default timeout for invalid levels.

Reply:
Added a check in `setFlushSoftTimeoutInSeconds` and `setFlushHardTimeoutsInMinutes`. Invalid levels are ignored and errors are printed to stderr.